### PR TITLE
some changes to the SDC Newton solver to make it more robust

### PR DESCRIPTION
**Pull request by zingale**
we now reject a solve if the mass fractions are too far out of [0, 1] also do the total energy update conservatively

## PR checklist

- [ ] test suite needs to be run on this PR
- [ ] this PR will change answers in the test suite to more than roundoff level
- [ ] all newly-added functions have docstrings as per the coding conventions
- [ ] the `CHANGES` file has been updated, if appropriate
- [ ] if appropriate, this change is described in the docs

### CODE REVIEW

**Comment by zingale** [comment]
test diffs are small:
http://groot.astro.sunysb.edu/Castro/test-suite/gfortran/2023-10-09-003/index.html

File: Source/sdc/sdc_newton_solve.H [hunk 1]
```diff
@@ -312,6 +322,20 @@ sdc_newton_subdivide(const Real dt_m,
...
+                // no point in continuing this subdivision if one stage failed
+                break;
+            }
+
+            // our solve may have resulted in mass fractions outside
+            // of [0, 1] -- reject if this is the case
+            for (int n = 0; n < NumSpec; ++n) {
+                if (U_new[UFS+n] < -newton::species_failure_tolerance * U_new[URHO] ||
+                    U_new[UFS+n] > (1.0_rt + newton::species_failure_tolerance) * U_new[URHO]) {
+                    ierr = newton::BAD_MASS_FRACTIONS;
```
Comment (maxpkatz) [inline Source/sdc/sdc_newton_solve.H:335]: Does it makes sense to keep doing the sub-stepping if we hit this error case?

**Comment by zingale** [comment]
`bubble_convergence` converges as expected for SDC-4:

```
 $\rho$                           3.585788e+15   3.265       3.729242e+14   3.727       2.816993e+13
 $\rho u$                         1.118655e+24   3.796       8.055982e+22   3.961       5.174599e+21
 $\rho v$                         1.312309e+24   3.544       1.124968e+23   3.852       7.792157e+21
 $\rho E$                          3.69012e+32   2.949       4.780079e+31   3.669        3.75798e+30
 $\rho e$                         3.689677e+32   2.949       4.779642e+31   3.669       3.757779e+30
 $T$                              1.437985e+18   3.508       1.264001e+17   3.831       8.884044e+15
 $\rho X(\isotm{He}{4})$          3.584552e+15   3.268       3.722151e+14   3.725       2.814812e+13
 $\rho X(\isotm{C}{12})$          1.519908e+13   2.544       2.605549e+12   3.797       1.874236e+11
 $\rho X(\isotm{O}{16})$              35854370   3.265            3729525   3.727           281671.7
 $\rho X(\isotm{Fe}{56})$             35856870   3.265            3729094   3.727           281699.3
```

File: Source/sdc/sdc_newton_solve.H (hunk 1 above)
Comment (maxpkatz) [reply Source/sdc/sdc_newton_solve.H:301]: Something like that, yes

Comment (zingale) [reply Source/sdc/sdc_newton_solve.H:301]: break added

Comment (maxpkatz) [reply Source/sdc/sdc_newton_solve.H:301]: Well I'm not as worried about a break inside the species loop, I mean a break after the species loop before we finish up anything else we do below this

Comment (zingale) [reply Source/sdc/sdc_newton_solve.H:301]: ah, okay

Comment (zingale) [reply Source/sdc/sdc_newton_solve.H:296]: I just moved it to before the previous break.  Thanks.  I missed that.

**Review by maxpkatz — APPROVED** [review]

**Pull request merged**